[Clang][AMDGPU] Accept builtins in lambda declarations

jmmartinez opened the pull request:
`Sema::getCurFunctionDecl(AllowLambda = false)` returns a nullptr when the lambda declaration is outside a function (for example, when assigning a lambda to a static constexpr variable).

This triggered an assertion in `SemaAMDGPU::CheckAMDGCNBuiltinFunctionCall`.

Using `Sema::getCurFunctionDecl(AllowLambda = true)` returns the declaration of the enclosing lambda.

Stumbled with this issue when refactoring some code in CK.

jmmartinez committed:
[Clang][AMDGPU] Accept builtins in lambda declarations

Sema::getCurFunctionDecl(AllowLambda = false) returns a nullptr when the
lambda declaration is outside a function (for example, when used in an
assignment).

Using Sema::getCurFunctionDecl(AllowLambda = true) returns the
declaration of the enclosing lambda.

jmmartinez commented on clang/test/SemaHIP/amdgpu-builtin-in-lambda.hip:
```diff
@@ -0,0 +1,24 @@
+        __builtin_amdgcn_global_load_lds(src, dst, 16, 0, 0); // expected-error{{invalid size value}} expected-note{{size must be 1, 2, or 4}}
+    };
+};
+
```
Add a test where we assign a value to a global outside a struct.

arsenm commented on clang/test/SemaHIP/amdgpu-builtin-in-lambda.hip:
```diff
@@ -0,0 +1,24 @@
+// REQUIRES: amdgpu-registered-target
```
Don't think it actually does

Artem-B replied:
I've just checked using experimental target `csky-unknown-elf ` that's not enabled by default, and clang indeed errors out if we attempt to generate code, but works OK with `-fsyntax-only`.

arsenm replied:
-emit-llvm should work fine, real codegen will obviously fail

shiltian commented on clang/lib/Sema/SemaAMDGPU.cpp:
```diff
@@ -27,7 +27,7 @@ bool SemaAMDGPU::CheckAMDGCNBuiltinFunctionCall(unsigned BuiltinID,
   unsigned OrderIndex, ScopeIndex;
 
-  const auto *FD = SemaRef.getCurFunctionDecl();
+  const auto *FD = SemaRef.getCurFunctionDecl(true);
```
```suggestion
  const auto *FD = SemaRef.getCurFunctionDecl(/*AllowLambda=*/true);
```

yxsamliu commented:
LGTM. Pls address other comments

arsenm reviewed (commented):
IIRC there were bugs when you try to use lambda in conjunction with the target attribute. Can you add a test where you try to use a builtin that depends on a subtarget feature, in a function marked with the target attribute (and not available in the main compile target), that we still error?

#60005 is the issue for OpenCL blocks, I assume lambdas have a similar issue

jmmartinez commented:
> IIRC there were bugs when you try to use lambda in conjunction with the target attribute. Can you add a test where you try to use a builtin that depends on a subtarget feature, in a function marked with the target attribute (and not available in the main compile target), that we still error?

I've added a separate test for this (the error is not triggered with -fsyntax-only but it is with -emit-llvm).

I've also added tests for local variables inside a function.
